client/app/bundles/course/level/pages/LevelsIndex/LevelsManager.tsx

```diff
@@ -0,0 +1,150 @@
+  const { t } = useTranslation();
+  const dispatch = useAppDispatch();
+
+  // Add 'deleted' property to each level in initial state
+  const [newLevels, setNewLevels] = useState(
+    levels.map((level) => ({ level, deleted: false })),
+  );
+  const [isSaving, setIsSaving] = useState(false);
+  const [isDirty, setIsDirty] = useState(false);
+
+  const handleAddLevel = (): void => {
+    setIsDirty(true);
+
+    const lastLevel = newLevels[newLevels.length - 1].level;
+    const newLevelThreshold = lastLevel ? lastLevel * 2 : 100;
+
+    const updatedLevels = [
+      ...newLevels,
+      { level: newLevelThreshold, deleted: false },
+    ];
+    setNewLevels(updatedLevels);
+  };
+
+  const handleThresholdChange = (index: number, newThreshold: string): void => {
+    const updatedLevels = [...newLevels];
+    updatedLevels[index + 1].level = parseInt(newThreshold, 10) || 0;
+
+    setNewLevels(updatedLevels);
+    setIsDirty(true);
+  };
+
+  const handleDeleteLevel = (index: number): void => {
+    const updatedLevels = [...newLevels];
+    updatedLevels[index + 1].deleted = true;
+    setNewLevels(updatedLevels);
+    setIsDirty(true);
+  };
+
+  const handleUndoDelete = (index: number): void => {
+    const updatedLevels = [...newLevels];
+    updatedLevels[index + 1].deleted = false;
+    setNewLevels(updatedLevels);
+    setIsDirty(true);
+  };
+
+  const handleSaveLevels = (): void => {
+    setIsSaving(true);
+    const levelsToSave = newLevels
+      .filter((level) => !level.deleted)
+      .map((level) => level.level);
+    dispatch(
+      saveLevels(
+        levelsToSave,
+        translations.saveSuccess,
+        translations.saveFailure,
+      ),
+    )
+      .then(() => {
+        setIsSaving(false);
+        setIsDirty(false);
+        setNewLevels(levelsToSave.map((level) => ({ level, deleted: false })));
+      })
+      .catch(() => {
+        setIsSaving(false);
+      });
+  };
+
+  const resetForm = (): void => {
+    setNewLevels(levels.map((level) => ({ level, deleted: false })));
+    setIsDirty(false);
+  };
+
+  return (
+    <Page
+      title={
+        <FormattedMessage {...defaultComponentTitles.course_levels_component} />
+      }
+    >
+      <LevelsTable
+        canManage={canManage}
+        isSaving={isSaving}
+        levels={newLevels}
+        onAddLevel={handleAddLevel}
+        onDeleteLevel={handleDeleteLevel}
+        onThresholdChange={handleThresholdChange}
+        onUndoDelete={handleUndoDelete}
+      />
+
+      <Slide direction="up" in={isDirty} unmountOnExit>
+        <div className="fixed inset-x-0 bottom-0 z-10 flex w-full items-center justify-between bg-neutral-800 px-8 py-4 text-white sm:bottom-8 sm:mx-auto sm:w-fit sm:rounded-lg sm:drop-shadow-xl">
+          <Typography>{t(translations.unsavedChanges)}</Typography>
+
+          <div className="ml-10">
+            <Button onClick={resetForm}>{t(translations.reset)}</Button>
+
+            <Button
```

Comment:
Have you considered my request last time, to have dialog when user is attempting to save the levels? If yes, may I know why you didn't add the dialog here?

I think it's important over here since the change actually might impact quite number of people (some might have their levels reduced without them knowing why, and some might have their levels suddenly bumped up). Might not be that shocking, but still we need to warn the user and assure them that the changes are the one they're going for

Reply:
I think levels are unlikely to be adjusted during a course, and if need be an instructor can always manually revert the changes. Adding an extra barrier to saving is unnecessary

Reply:
It's more towards the user experience. It might be not a frequent case that user adjusted the levels, but once it happens, we should be sure to do the right thing.

This is quite impactful, and hence we need to make user to be extra cautious while attempting to edit the levels.

Manually reverting changes is not a good advice; what if they accidentally click save after trying to delete 5 of those levels (and actually they still contemplating on deleting those)? Later when they wanted to revert the change, they need to re-remember what the entries of those 5 levels are, which are not good for users

Reply:
Does the dialog only appear if students levels are affected?  
i.e. Dialog does not appear if students levels remain unchanged (no students in course, or all students exp are such that they are unaffected by change, for example, all students exp is <100, and you're adding in new levels at >100)

Does the dialog show affected students?  
i.e. 5 students' levels will be affected by this change!

If dialog doesn't show these information, then don't bother implementing dialog.  
Also this page is not a priority, don't spend additional effort on it. Main objective is just to lift from js to ts.